### DISCUSSION TITLE
v2 动态增加删除 schema 目前是否无法实现，只能以默认隐藏的方式排除暂不需要的对象属性？

### DISCUSSION THREAD

**Discussion by userhjp** [alibaba/formily#1795] [Q&A 2.x] [answered]
例如：
```
const baseSchema = {
  type: 'object',
  properties: {
    userInfo: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          default: '张三',
        },
        age: {
          type: 'number',
          default: 18,
        },
      },
    },
  },
};
<SchemaField schema={baseSchema}></SchemaField>
```
这时表单属性只有 一个对象 userInfo 用户信息，如果我想给用户添加一辆车 car 并添加属性， 希望 schema变成如下：
```
const baseSchema = {
  type: 'object',
  properties: {
    userInfo: {
      type: 'object',
      properties: {
        name: {
          type: 'string',
          default: '张三',
        },
        age: {
          type: 'number',
          default: 18,
        },
      },
    },
    car: {
      type: 'object',
      properties: {
        color: {
          type: 'string',
          default: 'red',
        },
      },
    },
  },
};
```

也就动态添加删除schemaJSON，目前找了下api貌似没找到，还是没有这东西？，换个思路实现是将所有可以设置的扩展属性全部默认隐藏，需要添加的时候再显示出来，例如 car 属性对象设置默认隐藏 none，但是这么做如果在数据多的情况下会不会有问题，极端情况可能会有100+ 甚至200+ 个各种对象，会导致 JSONSchema 巨大，求大佬指点下思路，谢谢。

**Comment by janryWang** [comment, marked as answer]
看看designable-antd的配置面板实现

**Reply by userhjp** [reply]
好的，我研究研究 谢谢！

**Reply by userhjp** [reply]
尴尬了，以我目前的代码水平看 designable 源码着实比较吃力，找了半天愣是没看明白 组件 面板 表单是怎么关联起来的，拖动后怎么创建的没找到，找到了删除按钮，一路跟着引用找过去，想看怎么删除的，一层又一层。。一层又一层，看得一愣一愣的，(꒦_꒦) 留下来没技术的泪水 ，打扰了，告辞 
```
  removeNodes(nodes: TreeNode[]) {
    for (let i = nodes.length - 1; i >= 0; i--) {
      const node = nodes[i]
      if (node !== this.tree && node?.designerProps?.deletable !== false) {
        const previousIndex = node.index - 1
        const afterIndex = node.index + 1
        const parent = node.parent
        node.remove()
        const previous = previousIndex > -1 && parent.children[previousIndex]
        const after =
          afterIndex < parent.children.length && parent.children[afterIndex]
        this.selection.select(previous ? previous : after ? after : node.parent)
        this.hover.clear()
      }
    }
  }
```

**Reply by darkmice** [reply]
https://github.com/alibaba/formily/blob/5f8c187959c4b77028a167eb8e2e4e066cad6556/designable/antd/src/components/DesignableField/index.tsx#L26

看这里实现

**Reply by userhjp** [reply]
(⊙o⊙)…  感谢！

**Comment by janryWang** [comment]
其实很简单的，就是createForm的时候传入的values保证是一个observable数据，然后createForm在useMemo里用，useMemo的dependencies依赖observable数据的某个字段，字段发生变化会触发外层组件重新渲染，又会触发useMemo自动刷新form实例，这就达到了表单内某个字段控制整个schema片段动态更新的目的

**Reply by janryWang** [reply]
需要保证外层组件是observer包装过的，总之一个思路就是让外层组件的渲染受控于表单的值，然后在外层组件渲染的过程动态切换schema

**Reply by userhjp** [reply]
这么说就好理解多了，懂了，谢了!

**Reply by userhjp** [reply]
大佬，再请教下，form实例刷新后会让表单重新渲染，并且已存在的状态丢失，是需要要自己维护表单字段状态么，例如，原有字段 `{ a : '1 }' `添加了个字段 `{ b : '' }`  刷新form 后 `{ a :  '', b :  '' } ` a状态丢失，这里只能在表单schema修改前先getValues 保存一份状态，然后添加schema后再交给form更新时赋值么？是我思路不对还是怎么的，总感觉哪里不对劲的样子。

**Reply by janryWang** [reply]
传给form的values如果是observable数据，表单状态失效对它不影响的

**Reply by userhjp** [reply]
貌似想复杂了，我这样已经实现了，可能有点low - -，但是功能实现了，动态添加schema，form实例在父组件创建后就没管了， SchemaField 在子组件，schema依赖于dva，通过dva共享并操作schema添加删除会触发SchemaField所在组件更新，表单原有状态没有丢失，不知道这用法有没有什么后遗症 - -，先实现功能后期慢慢优化 了 。
```
const { schema } = useSelector((state: any) => state.visual as visualModelState);
<SchemaField schema={schema} scope={{ formTab }}></SchemaField>
```

**Reply by janryWang** [reply]
都用formily2.0了，还在用dva？reactive一把梭

**Reply by userhjp** [reply]
- - ！ 好吧，我这就去学
